src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,199 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node first;
+    private Node last;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        Node node = new Node<>(null, value, null);
+        if (size == 0) {
+            first = node;
+            last = node;
+        } else {
+            last.next = node;
+            node.prev = last;
+            last = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        Node node = new Node<>(null, value, null);
+
+        if (index == 0) {
+            if (size == 0) {
+                first = last = node;
+            } else {
+                node.next = first;
+                first.prev = node;
+                first = node;
+            }
+        } else if (index == size) {
+            last.next = node;
+            node.prev = last;
+            last = node;
+        } else {
+            Node n = first;
+            for (int i = 0; i < index - 1; i++) {
+                n = n.next;
+            }
+            node.next = n.next;
+            node.prev = n;
+            n.next.prev = node;
+            n.next = node;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.size() == 0) {
+            return;
+        }
+
+        for (int i = 0; i < list.size(); i++) {
+            Node newNode = new Node<>(last, list.get(i), null);
```

Comment:
Ensure that the `Node` instantiation uses the generic type `T` for type safety, i.e., `Node<T> newNode = new Node<>(last, list.get(i), null);`.